File: algorithms/2_medium/M/02134_minimum_swaps_to_group_all_1s_together_ii/sliding_window.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static int minSwaps(const vector<int>& nums) {
        const isize len = std::ssize(nums);
        isize ones = 0;

        for (isize i = 0; i < len; i += 1) {
            assert(nums[i] == 0 || nums[i] == 1);
            ones += nums[i];
        }

        if (ones <= 1 || (len - ones) <= 1) {
            return 0;
        }

        isize freq_window = 0;
        for (isize i = len - ones; i < len; i += 1) { freq_window += nums[i]; }

        isize need = ones - freq_window;

        for (isize i = 0; i < len - 1; i += 1) {
            freq_window += nums[i];
            freq_window -= nums[(len - ones + i) % len];
            need = std::min(need, ones - freq_window);
        }

        return static_cast<int>(need);
    }
};

File: algorithms/2_medium/M/02134_minimum_swaps_to_group_all_1s_together_ii/sliding_window.rs
impl Solution {
    pub fn min_swaps(nums: Vec<i32>) -> i32 {
        let len = nums.len();
        let ones = nums.iter().filter(|&&x| x == 1).count();
        
        if ones <= 1 || (len - ones) <= 1 {
            return 0;
        }

        let mut freq_window = nums[(len - ones)..len].iter().filter(|&&x| x == 1).count();
        let mut need = ones - freq_window;

        for i in 0..(len - 1) {
            if nums[i] == 1 {
                freq_window += 1;
            }

            if nums[(len - ones + i) % len] == 1 {
                freq_window -= 1;
            }

            need = need.min(ones - freq_window);
        }

        need.try_into().unwrap()
    }
}
